### PR TITLE
[CNSMR-2303] Add documentation for Danger rules

### PR DESCRIPTION
**Pull request by AliSoftware**
This introduces a document describing all the current rules we wrote in Danger, what they do, how they work, what they mean and how to react when you trigger them.

### CODE REVIEW

File: Cookbook/Technical-Documents/DangerRules.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+This rule is only a warning because we don't want it to block PRs, as there are some rare exceptions where we allow the limit to be exceeded, as long as it's justified by the author (e.g. a refactoring PR)
+
+## Podfile Checks
+
+> Declared in: `danger-ci/pods.rb`
+
+### Check that Podfile.lock has been properly commited
+
+> Function: `check_podfile_lock_commited`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Function: `check_podfile_lock_committed`  
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+_Note: We have an informal limit of 800 lines for PRs, but since it's only informal, we decided to only make Danger put a message if we're more then 50 lines over that informal limit._
+
+This rule is only a warning because we don't want it to block PRs, as there are some rare exceptions where we allow the limit to be exceeded, as long as it's justified by the author (e.g. a refactoring PR)
+
+## Podfile Checks
+
+> Declared in: `danger-ci/pods.rb`
+
+### Check that Podfile.lock has been properly commited
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
### Check that Podfile.lock has been properly committed
```

**Review by AliSoftware — COMMENTED** [review]
typos

File: Cookbook/Technical-Documents/DangerRules.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# Danger Configuration
+
+We've configured [Danger](https://danger.systems/ruby/) in our main repository to help us prevent some mistakes we saw happening in the past and to assis us during code review.
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
We've configured [Danger](https://danger.systems/ruby/) in our main repository to help us prevent some mistakes we saw happening in the past and to assist us during code review.
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+This document aims to document the rules we added in our `Dangerfile`.
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
The purpose of this file is to document all the rules we wrote in our `Dangerfile`, what they mean and how to react to them when they are triggered.
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+2. Remind you to **update the `SDK/CHANGELOG.md` document** (intended for our partners) **when you make a significant change to the SDK**
+  - If you made trivial changes to the SDK which you consider don't require to be mentioned in the CHANGELOG for partners, you can add the `#IgnoreSDKChanges` in the PR title to acknowledge that it's ok that the `SDK/CHANGELOG.md` wasn't updated. **When in doubt if a change in SDK needs to be mentioned in the CHANGELOG, ask the `#sdk_squad` in slack**.
+
+
+Context | Mention in PR title | Commit included in SDK CRP ticket? | New entry in CHANGELOG?
+---|---|---|---
+This PR didn't touch any (non-test) file in `sdk/*`, nothing related to SDK here | _None_ | 🚫 No | 🚫 No
+This PR made significant SDK changes | `#SDK` (or `[SDK-xxx]`) | ✅ Yes | ✅Yes
+This PR touched `sdk/*` files but it's a false positive (e.g. we just fixed indentation or a typo in a comment) |`#IgnoreSDKChanges` |  🚫 No | 🚫 No
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
This PR touched `sdk/*` files but it's a false positive (e.g. we just fixed indentation or a typo in a comment) | `#IgnoreSDKChanges` | 🚫 No | 🚫 No
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+This PR made SDK changes (that need to appear in CRP) but they are not worth a note in our partner's CHANGELOG | `#SDK ` (or `[SDK-xxx]`) AND `#IgnoreSDKChanges` | ✅ Yes | 🚫 No
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
This PR made SDK changes (which need to appear in CRP) but they are not worth a note in our partner's CHANGELOG | `#SDK ` (or `[SDK-xxx]`) *AND* `#IgnoreSDKChanges` | ✅ Yes | 🚫 No
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+> Function: `warn_sdk_release_changelog`  
+> Type: ⚠️ warning
+
+This rule ensures that, when we do a **PR to make a public SDK release** (i.e. to merge a `release/sdk/*` branch), **the `SDK/CHANGELOG.md` has been updated**.
+
+Especially that in the top section title we **replace `## x.x.x (TBD)` with the actual version number and release date**.
+
+## pbxproj checks
+
+> Declared in: `danger-ci/xcodeproj_utils.rb`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Declared in: `danger-ci/pbxproj.rb`
```
(renamed in https://github.com/Babylonpartners/babylon-ios/pull/8368)

File: Cookbook/Technical-Documents/DangerRules.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+This rule is to alert us if we **forgot to commit the changes made in `Podfile.lock`**. This can happen if:
+
+ - We modified the `Podfile` but forgot to do `bundle exec pod install` afterwards
+ - We modified the `Podfile` and ran `pod install` but forgot to include the `Podfile.lock` in the commit
+ - We modified the `Podfile` once, ran `pod install` and commited the `Podfile.lock` changes for that first modification in one commit, but then later re-modified the `Podfile` and forgot to commit the second change to `Podfile.lock` (which means that the PR will contain changes in `Podfile.lock`, but not the expected ones)
+
+### Inform if dependencies might have change
+
+> Function: `warn_dependencies_changed`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Function: `check_dependencies_changed`  
```

(Renamed in https://github.com/Babylonpartners/babylon-ios/pull/8368)

File: Cookbook/Technical-Documents/DangerRules.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+The goal is to suggest them to take extra care in reviewing why we needed those changes in our dependencies and if they were properly justified.
+
+## SDK Checks
+
+> _Declared in: `danger-ci/sdk.rb`_
+
+### Mark changes to SDK with `#SDK` (or `#IgnoreSDKChanges`)
+
+> Function: `warn_sdk_hashtag`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Function: `check_sdk_hashtag`  
```

(Renamed in https://github.com/Babylonpartners/babylon-ios/pull/8368)

File: Cookbook/Technical-Documents/DangerRules.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+Context | Mention in PR title | Commit included in SDK CRP ticket? | New entry in CHANGELOG?
+---|---|---|---
+This PR didn't touch any (non-test) file in `sdk/*`, nothing related to SDK here | _None_ | 🚫 No | 🚫 No
+This PR made significant SDK changes | `#SDK` (or `[SDK-xxx]`) | ✅ Yes | ✅Yes
+This PR touched `sdk/*` files but it's a false positive (e.g. we just fixed indentation or a typo in a comment) |`#IgnoreSDKChanges` |  🚫 No | 🚫 No
+This PR made SDK changes (that need to appear in CRP) but they are not worth a note in our partner's CHANGELOG | `#SDK ` (or `[SDK-xxx]`) AND `#IgnoreSDKChanges` | ✅ Yes | 🚫 No
+
+### Ensure SDK CHANGELOG is updated on PRs releasing the SDK
+
+> Function: `warn_sdk_release_changelog`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Function: `check_sdk_release_changelog`  
```

(Renamed in https://github.com/Babylonpartners/babylon-ios/pull/8368)

File: Cookbook/Technical-Documents/DangerRules.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+Especially that in the top section title we **replace `## x.x.x (TBD)` with the actual version number and release date**.
+
+## pbxproj checks
+
+> Declared in: `danger-ci/xcodeproj_utils.rb`
+
+### Warn for unexpected resources in targets
+
+> Function: `warn_for_unexpected_resource_files`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Function: `check_unexpected_resource_files`  
```

(Renamed in https://github.com/Babylonpartners/babylon-ios/pull/8368)

File: Cookbook/Technical-Documents/DangerRules.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+> Function: `warn_for_unexpected_resource_files`  
+> Type: 🚫 failure
+
+This rule ensure that we didn't accidentally add some Xcode-specific files to a target's "Copy Resource Files" phase, as those files are only used by Xcode or for documentation or tests, but that shouldn't end up in the final framework or app.
+
+This rule currently checks for `*.xcconfig`, `*.md`, `*Info.plist` and any file under a `__Snapshots__` directory and fail if they have been added to your target accidentally.
+
+### Warn about `null` references
+
+> Function: `warn_null_references`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Function: `check_null_references`  
```

(Renamed in https://github.com/Babylonpartners/babylon-ios/pull/8368)

File: Cookbook/Technical-Documents/DangerRules.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+During a merge of Xcode project gone bad, `(null)` references can appear in the `.pbxproj` file.
+
+This rule will trigger when any `(null)` references or `Recovered References` got introduced in the `pbxproj` that your PR touched. Additionally, it will also add inline comments to pinpoint the lines in your `pbxproj` where those null references were found.
+
+When that happens, we need to understand what went wrong during the `pbxproj` merge and fix it properly (and manually) before we can merge the PR.
+
+### Warn about orphan UUIDs
+
+> Function: `warn_orphan_uuids`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Function: `check_orphan_uuids`  
```

(Renamed in https://github.com/Babylonpartners/babylon-ios/pull/8368)

File: Cookbook/Technical-Documents/DangerRules.md [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+If you plan to add a new Danger rule:
+
+ * Please follow the naming convention of starting the functions implementing the rules with `check_`
+ * Make sure you keep the main `Dangerfile` clean: the main file is supposed to only call `check_*` functions defined in other files, not to contain actual rule implementations
+ * Make sure you document the new rule here
+
+## PR Size
+
+> Declared in: `danger-ci/pr_size.rb`
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
> Declared in: `danger-ci/pr_size.rb`  
> Function: `check_pr_size`
```

**Review by AliSoftware — COMMENTED** [review]
Proof-reading + update function names after https://github.com/Babylonpartners/babylon-ios/pull/8368 's approval

**Comment by AliSoftware** [comment]
Related PR adding a Danger rule in main repo to remind to update this docs here (so that it doesn't become out of sync): https://github.com/Babylonpartners/babylon-ios/pull/8386

File: Cookbook/Technical-Documents/DangerRules.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+### Check that Podfile.lock has been properly committed
+
+> * Function: `check_podfile_lock_committed`
+> * Type: 🚫 failure
+
+This rule is to alert us if we **forgot to commit the changes made in `Podfile.lock`**. This can happen if:
+
+ - We modified the `Podfile` but forgot to do `bundle exec pod install` afterwards
+ - We modified the `Podfile` and ran `pod install` but forgot to include the `Podfile.lock` in the commit
+ - We modified the `Podfile` once, ran `pod install` and commited the `Podfile.lock` changes for that first modification in one commit, but then later re-modified the `Podfile` and forgot to commit the second change to `Podfile.lock` (which means that the PR will contain changes in `Podfile.lock`, but not the expected ones)
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
 - We modified the `Podfile` once, ran `pod install` and committed the `Podfile.lock` changes for that first modification in one commit… but then later re-modified the `Podfile` and forgot to commit the second change to `Podfile.lock` (which means that the PR will contain changes in `Podfile.lock`, but not the expected ones)
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+> * Function: `check_podfile_lock_committed`
+> * Type: 🚫 failure
+
+This rule is to alert us if we **forgot to commit the changes made in `Podfile.lock`**. This can happen if:
+
+ - We modified the `Podfile` but forgot to do `bundle exec pod install` afterwards
+ - We modified the `Podfile` and ran `pod install` but forgot to include the `Podfile.lock` in the commit
+ - We modified the `Podfile` once, ran `pod install` and committed the `Podfile.lock` changes for that first modification in one commit… but then later re-modified the `Podfile` and forgot to commit the second change to `Podfile.lock` (which means that the PR will contain changes in `Podfile.lock`, but not the expected ones)
+
+### Inform if dependencies might have change
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
### Inform if dependencies might have changed
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+This rule will trigger when any `(null)` references or `Recovered References` got introduced in the `pbxproj` that your PR touched. Additionally, it will also add inline comments to pinpoint the lines in your `pbxproj` where those null references were found.
+
+When that happens, we need to understand what went wrong during the `pbxproj` merge and fix it properly (and manually) before we can merge the PR.
+
+### Warn about orphan UUIDs
+
+> * Function: `check_orphan_uuids`
+> * Type: 🚫 failure (+ inline comments as warnings)
+
+During a merge of Xcode project gone bad, orphan UUDS can appear in the `.pbxproj` file.
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
During a merge of Xcode project gone bad, orphan UUIDs can appear in the `.pbxproj` file.
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+
+_Context: The `pbxproj` format is such that it contains a list of file references (with an associated UUID) listing all the files that are in the pbxproj, then points to those references in the various parts of the `pbxproj` (for example in the "Compile Sources" build phase part, to reference which files are included in that build phase). If a `pbxproj` merge went wrong, sometimes the line declaring the file reference (and its UUID) could be deleted during the bad merge while other lines in the pbxproj referencing that UUID would still remain._
+
+This rule will warn you if any part of the pbxproj reference an UUID that got accidentally removed. (Note: you should see similar warnings when you run `pod install` in your terminal and it warns about an unknown UUID)
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
This rule will warn you if any part of the pbxproj references an UUID that got accidentally removed. (Note: you should see similar warnings when you run `pod install` in your terminal and it warns about an unknown UUID)
```

File: Cookbook/Technical-Documents/DangerRules.md [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+
+When that happens, you need to figure out if the removal of the file was intended as part of your PR (and if so, fix trhe merge by also removing the lines pointing to that now-deleted file reference), or if the removal of the file from the pbxproj was unintented (and if so, restore it). You could do that by editing the `pbxproj` manually, but if possible it's even better if you can fix it in Xcode (e.g. by removing the offending file from the project and add it again)
```
Comment (AliSoftware) [inline Cookbook/Technical-Documents/DangerRules.md]:
```suggestion
When that happens, you need to figure out if the removal of the file was intended as part of your PR (and if so, fix the merge by also removing the lines pointing to that now-deleted file reference), or if the removal of the file from the pbxproj was unintented (and if so, restore it). You could do that by editing the `pbxproj` manually, but if possible it's even better if you can fix it in Xcode (e.g. by removing the offending file from the project and add it again)
```

**Review by viorel15 — APPROVED** [review]

**Review by MichalTKwiecien — APPROVED** [review]

**Pull request merged**